File: json-gen/py/osc.py
import json
from itertools import chain

def opt_items(x):
    name, value = x
    if value:
        return [(name, value)]
    else:
        return []

def dict_opt(d, rest):
    lst = map(opt_items, rest.items())
    xs = list(chain.from_iterable(lst))
    return dict(d.items() + xs)

def root(windows, keys = [], init_send = []): 
    return { "main": windows, "keys": keys, "init-send": init_send }

def window(content, title = "", size = None, keys = []):    
    return dict_opt({ "content": content, "keys": keys, "title": title }, {"size", size})

def ui(sym, id = None, send = None):
    return dict_opt(sym, {"id": id, "send": send})

def keys(xs):
    return { "keys": xs }

def key(name, msgs):
    return { "key": msgs }

def msg(client, path, args):
    return { "client": client, "path": path, "args": args }

def ref(n):
    return "$" + str(n)

def send(default = [], cases = [], cases_off = []):
    def make_case(x):
        return ("case " + x[0], x[1])

    def make_case_off(x):
        return ("case-off " + x[0], x[1])      

    return dict([("default", default)] + map(make_case, cases) + map(make_case_off, cases))

# layout widgets

def hor(items):
    return { "hor": items }

def ver(items):
    return { "ver": items }

def space(size):
    return { "space": size }

def tabs(pages):
    return { "tabs": pages }

def page(content, title = ""):
    return { "page": { "content": content, "title": title }}

# valuators

def from_pair_range(range):
    if range:
        return [range[0], range[1]]
    else:
        return None    

def dial(init = None, color = None, range = None):
    return { "dial": dict_opt({}, { "init": init, "color": color, "range": from_pair_range(range) })}

def hfader(init = None, color = None, range = None):
    return { "hfader": dict_opt({}, { "init": init, "color": color, "range": from_pair_range(range) })}

def vfader(init = None, color = None, range = None):
    return { "vfader": dict_opt({}, { "init": init, "color": color, "range": from_pair_range(range) })}

def toggle(init = None, color= None, text = None):
    return { "toggle": dict_opt({}, { "init": init, "color": color, "text": text })}

def button(color = None, text = None):
    return { "button": dict_opt({}, { "color": color, "text": text })}

def circle_toggle(init = None, color = None):
    return { "circle-toggle": dict_opt({}, { "init": init, "color": color, "text": text })}

def circle_button(color = None):
    return { "circle-button": dict_opt({}, { "color": color, "text": text })}

def int_dial(init = None, color = None, range = None):
    return { "int-dial": dict_opt({}, { "init": init, "color": color, "range": from_pair_range(range) })}

def label(color = None, text = None):
    return { "label": dict_opt({}, { "color": color, "text": text })}

def hcheck(init = None, color = None, size = None, texts = None, allow_deselect = None):
    return { "hcheck": dict_opt({}, { "init": init, "color": color, "size": size, "texts": texts, "allow-deselect": allow_deselect })}

def vcheck(init = None, color = None, size = None, texts = None, allow_deselect = None):
    return { "vcheck": dict_opt({}, { "init": init, "color": color, "size": size, "texts": texts, "allow-deselect": allow_deselect })}

def xypad(init = None, color = None):
    return { "xy-pad": dict_opt({}, { "init": from_pair_range(init), "color": color }) }

def multi_toggle(init = [], size = None, color = None, texts = []):
    return { "multi-toggle": dict_opt({ "texts": texts }, { "init": init, "size": from_pair_range(size), "color": color }) }

def hfader_range(init = None, color = None):
    return { "hfader-range": dict_opt({}, { "init": from_pair_range(init), "color": color }) }

def vfader_range(init = None, color = None):
    return { "vfader_range": dict_opt({}, { "init": from_pair_range(init), "color": color }) }    

def xypad_range(init_x = None, init_y = None, color = None):
    return { "xy-pad-range": dict_opt({}, { "init-x": from_pair_range(init_x), "init-y": from_pair_range(init_y), "color": color })}

def drop_down_list(init = None, texts = []):
    return { "drop-down-list": dict_opt({ "texts": texts }, { "init": init}) }

def text_input(init = None, color = None, text_length = None):
    return { "text-input": dict_opt({}, { "init": init, "color": color, "text-length": text_length }) }

def file_input(init = None, color = None, text = None):
    return { "file-input": dict_opt({}, { "init": init, "color": color, "text": text }) }

def double_check(init = None, sizes = [], color1 = None, color2 = None, texts = [], orient = None, allow_deselect = None):
    return { "double-check": dict_opt({"texts": texts, "sizes": sizes}, { "init": from_pair_range(init) }), "color1": color1, "color2": color2, "orient": orient, "allow-deselect": allow_deselect }

def orient(is_first = True, is_first_hor = True, is_second_hor = True):
    return [is_first, is_first_hor, is_second_hor]

def write_json(filename, data):
    with open(filename, 'w') as outfile:
        json.dump(data, outfile)

File: json-gen/hs/dragon-osc/src/Dragon/Osc.hs
{-# Language OverloadedStrings #-}
module Dragon.Osc(
    Size, Float2, Int2,
    Root(..), Window(..), Ui(..), Param(..), 
    Keys, KeyEvent(..), 
    HotKey(..), withModifiers, ctrl, shift, meta, alt,
    Send(..), Msg(..), Args, Arg(..), 
    Page(..), Sym(..), Orient(..),
    setId, setSend, setMsgs, setMsg, ui, sendMsg, onBool, onBools, 
    multiUi,
    writeJson
) where

import Data.Aeson
import Data.String
import Control.Arrow(first, second)
import qualified Data.Map as M
import Data.Monoid
import Data.Maybe

import qualified Data.ByteString.Lazy as LB

type Size = (Int, Int)
type Float2 = (Float, Float)
type Int2 = (Int, Int)

data Root = Root 
    { rootWindows :: [Window] 
    , rootKeys    :: Keys 
    , rootInitOsc :: [Msg] }

data Window = Window 
    { windowTitle   :: String
    , windowSize    :: Maybe Size    
    , windowContent :: Ui
    , windowKeys    :: Keys }

data Ui = Ui 
    { uiSym     :: Sym
    , uiParam   :: Param  }

data Param = Param 
    { paramId   :: Maybe String
    , paramSend :: Maybe Send }

type Keys = [KeyEvent]
data KeyEvent = KeyEvent { key:: HotKey, send:: Send }
newtype HotKey = HotKey { unHotKey :: [String] }

instance IsString HotKey where
    fromString key = HotKey [key]

withModifiers :: [String] -> HotKey -> HotKey
withModifiers xs (HotKey ys) = HotKey (ys ++ xs)

ctrl, shift, meta, alt :: HotKey -> HotKey

ctrl  = withModifiers ["ctrl"]
shift = withModifiers ["shift"]
meta  = withModifiers ["meta"]
alt   = withModifiers ["alt"]

data Send = Send 
    { sendDefault :: [Msg]
    , onValue :: [(String, [Msg])]
    , onValueOff :: [(String, [Msg])] }

data Msg = Msg 
        { msgClient :: String
        , msgPath   :: String
        , msgArgs   :: Args }
    | DelayedMsg
        { msgClient :: String
        , msgPath   :: String
        , msgArgs   :: Args
        , msgDelay  :: Float }

type Args = [Arg]

data Arg = ArgString String | ArgFloat Float | ArgBool Bool | ArgInt Int | Arg Int | Mem String

data Page = Page 
    { pageTitle :: String
    , pageContent:: Ui
    , pageKeys :: Keys }

data Sym 
    = Hor [Ui]
    | Ver [Ui]

    | Tabs [Page]
    | Space
    | Glue

    | Dial   { init:: Float, color:: String, range:: Float2 }
    | HFader { init:: Float, color:: String, range:: Float2 }
    | VFader { init:: Float, color:: String, range:: Float2 }

    | Toggle { initBool  :: Bool, color:: String, text:: String }
    | IntDial { initBool :: Bool, color:: String, rangeInt :: Int2 }
    | Button { color:: String, text:: String }
    | Label  { color:: String, text:: String }

    | CircleToggle { initBool :: Bool, color :: String }
    | CircleButton { color :: String }

    | MultiToggle { initSet :: [Int2], size:: Size, color:: String, texts:: [String] }
    | HCheck      { initInt :: Int,   leng:: Int, color:: String, texts:: [String], allowDeselect :: Maybe Bool }
    | VCheck      { initInt :: Int,   leng:: Int, color:: String, texts:: [String], allowDeselect :: Maybe Bool }

    | XYPad { initFlow2:: Float2, color:: String }
    | HFaderRange { initFlow2:: Float2, color:: String }
    | VFaderRange { initFlow2:: Float2, color:: String }
    | XYPadRange { initX:: Float2, initY:: Float2, color:: String }

    | DropDownList { initInt:: Int, texts:: [String] }
    | TextInput { initString:: Maybe String, color:: String, textLength :: Maybe Int }    
    | FileInput { initString:: Maybe String, color:: String, text:: String }
    | DoubleCheck { initInt2:: Int2, sizes:: [Int], color1:: String, color2:: String, doubleTexts:: [(String, [String])], orient:: Orient, allowDeselect:: Maybe Bool }

data Orient = Orient { orientIsFirst :: Bool, orientIsFirstHor :: Bool, orientIsSecondHor :: Bool }

-------------------------------------------------------------
-- aux

setId :: String -> Ui -> Ui
setId name (Ui sym param) = Ui sym (param { paramId = Just name })

setSend :: Send -> Ui -> Ui
setSend send (Ui sym param) = Ui sym (param { paramSend = Just send })

setMsgs :: [Msg] -> Ui -> Ui
setMsgs = setSend . (\msgs -> Send msgs [] [])

setMsg :: Msg -> Ui -> Ui
setMsg = setMsgs . return

ui x = Ui x (Param Nothing Nothing)

sendMsg :: Msg -> Send
sendMsg msg = Send [msg] [] []

onBool :: Msg -> Msg ->  Ui -> Ui
onBool on off = onBools [on] [off]

onBools :: [Msg] -> [Msg] ->  Ui -> Ui
onBools ons offs = setSend (Send [] onVal []) 
    where onVal = [("true", ons), ("false", offs)]

-------------------

multiUi :: (Int, Int) -> (Int -> Ui) -> Ui
multiUi (xSize, ySize) unit = ui $ Ver $ fmap row $ take ySize [0, xSize .. ]
    where
        row n = ui $ Hor $ fmap (unit . (+ n)) [0 .. xSize - 1]

-------------------------------------------------------------
-- toJson

instance ToJSON Root where
    toJSON (Root windows keys initSend) = object ["main" .= windows, "keys" .= keys, "init-send" .= initSend]

instance ToJSON Window where
    toJSON (Window title size content keys) = "window" =: (object $ catMaybes [Just $ "title" .= title, fmap (\sz -> "size" .=  [fst sz, snd sz]) size, Just $ "content" .= content, Just $ "keys" .= keys])

instance ToJSON Orient where
    toJSON (Orient isFirst isFirstHor isSecondHor) = toJSON $ [isFirst, isFirstHor, isSecondHor]

instance ToJSON Ui where
    toJSON (Ui sym param) = case toJSON sym of
        Object obj -> case toJSON param of
            Object paramObj -> Object (obj <> paramObj)
            _               -> Object obj
        value -> value

instance ToJSON Param where
    toJSON (Param id send) = object $ catMaybes [fmap ("id" .= ) id, fmap ("send" .= ) send]

(=:) :: (ToJSON a) => String -> a -> Value
(=:) name value = toJSON $ M.fromList [(name, value)]

instance ToJSON Page where
    toJSON p = "page" =: object ["title" .= pageTitle p, "content" .= pageContent p, "keys" .= pageKeys p ] 

instance ToJSON KeyEvent where
    toJSON k = object ["key" .= key k, "send" .=  send k ]
        where 
            

instance ToJSON HotKey where
    toJSON (HotKey xs) = case xs of
        [k] -> toJSON k
        _   -> toJSON xs

instance ToJSON Send where
    toJSON (Send defaults onValue onValueOff) = object $ ("default" .= defaults) : (tfm "case" onValue ++ tfm "case-off" onValueOff)
            where tfm name m = fmap (uncurry (.=)) $ fmap (first (fromString . ( (name ++ " ") ++ ))) m

instance ToJSON Msg where
    toJSON (Msg client path args) = "msg" =: object ["client" .= client, "path" .= path, "args" .= args]

    toJSON (DelayedMsg client path args delay) = "msg" =: object ["client" .= client, "path" .= path, "args" .= args, "delay" .= delay]


instance ToJSON Arg where
    toJSON a = case a of
        ArgString x -> toJSON x
        ArgFloat x  -> toJSON x 
        ArgBool x -> toJSON x 
        ArgInt x -> toJSON x
        Arg n -> toJSON $ '$' : show n
        Mem name -> toJSON $ '$' : name

instance ToJSON Sym where
    toJSON x = case x of
        Hor xs -> "hor" =: xs
        Ver xs -> "ver" =: xs

        Tabs ps -> "tabs" =: ps
        Space   -> "space" =: object []
        Glue    -> "glue"  =: object []

        Dial init color range   -> floatVal "dial"   init color range
        HFader init color range -> floatVal "hfader" init color range
        VFader init color range -> floatVal "vfader" init color range

        Toggle init color text -> "toggle" =: object [ "init" .= init, "color" .= color, "text" .= text ]
        IntDial init color range -> "int-dial" =: object [ "init" .= init, "color" .= color, "range" .= fromPair range ]
        Button color text -> "button" =: object [ "color" .= color, "text" .= text ]
        Label color text -> "label" =: object [ "color" .= color, "text" .= text ]

        CircleToggle init color -> "circle-toggle" =: object [ "init" .= init, "color" .= color ]
        CircleButton color -> "circle-button" =: object [ "color" .= color ]

        MultiToggle  initSet size color texts -> "multi-toggle" =: object [ "init" .= initSet, "color" .= color, "texts" .= texts, "size" .= fromPair size ]
        HCheck init leng color texts allowDeselect -> "hcheck" =: (object $ (let xs = [ "init" .= init, "size" .= leng, "color" .= color, "texts" .= texts ] in maybe xs (:xs) (fmap ("allow-deselect" .=) allowDeselect)))
        VCheck init leng color texts allowDeselect -> "vcheck" =: (object $ (let xs = [ "init" .= init, "size" .= leng, "color" .= color, "texts" .= texts ] in maybe xs (:xs) (fmap ("allow-deselect" .=) allowDeselect)))

        XYPad (initX, initY) color -> "xy-pad" =: object [ "init" .= [initX, initY], "color" .= color ]

        DropDownList init texts -> "drop-down-list" =: object [ "init" .= init, "texts" .= texts ]
        TextInput maybeInit color textLength -> "text-input" =: (object $ catMaybes [fmap ("init" .= ) maybeInit, Just $ "color" .= color, fmap ("text-length" .= ) textLength])
        FileInput maybeInit color text -> "file-input" =: (object $ catMaybes [fmap ("init" .= ) maybeInit, Just $ "color" .= color, Just $ "text" .= text  ])    
        DoubleCheck initInt2 sizes color1 color2 doubleTexts orient allowDeselect -> "double-check" =: object [ "init" .= [fst initInt2, snd initInt2], "sizes" .= sizes, "color1" .= color1, "color2" .= color2, "texts" .= doubleTexts, "orient" .= orient ]
        where 
            floatVal name init color (rangeMin, rangeMax) = name =: object [ "init" .= init, "color" .= color, "range" .= [rangeMin, rangeMax] ]
            fromPair (a, b) = [a, b]

-----------------------------------------------------------------

writeJson :: String -> Root -> IO ()
writeJson filename root = do
    LB.writeFile "test.json" $ encode root
